File: processor/src/main/java/de/qualityminds/lazyval/processor/codegen/JpaGenerator.java
package de.qualityminds.lazyval.processor.codegen;

import com.palantir.javapoet.*;
import de.qualityminds.lazyval.processor.spi.FilePerTypeGenerator;
import de.qualityminds.lazyval.processor.spi.GeneratorResult;
import de.qualityminds.lazyval.processor.spi.ValidatedGeneratorElement;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import java.util.Collection;
import java.util.List;
import java.util.Set;

// tag::docu[]
public class JpaGenerator implements FilePerTypeGenerator {

    private static final String GENERATOR_ID = "jpa";
    private static final String OPTION_GENERATED_PACKAGE = "lazyval.jpa.generatedPackage";

    @Override
    public String generatorId() {
        return GENERATOR_ID;
    }

    @Override
    public Collection<String> requiredClasspath() {
        return List.of("jakarta.persistence.AttributeConverter");
    }

    @Override
    public Set<String> supportedOptions() {
        return Set.of(OPTION_GENERATED_PACKAGE);
    }

    public GeneratorResult generateFilePerType(ValidatedGeneratorElement validatedElement, Settings userSettings){
        // end::docu[]
        TypeElement element = validatedElement.element();

        TypeMirror type = element.asType();
        TypeMirror wrappedType = validatedElement.wrappedType();
        TypeName wrappedTypeName;
        if (wrappedType.getKind().isPrimitive()) {
            // Box primitive types for JPA generics
            wrappedTypeName = TypeName.get(wrappedType).box();
        } else {
            wrappedTypeName = TypeName.get(wrappedType);
        }

        // Jpa Converter needs to check for null due to boxing primitive types
        final MethodSpec convertToDatabaseColumn = MethodSpec.methodBuilder("convertToDatabaseColumn")
                    .addModifiers(Modifier.PUBLIC)
                    .returns(wrappedTypeName)
                    .addParameter(TypeName.get(type), "type")
                    .beginControlFlow("if(type == null)")
                    .addStatement("return null")
                    .endControlFlow()
                    .addStatement(String.format("return type.%s()", validatedElement.wrappedTypeName()))
                    .build();

        var lazyvalTypeName = TypeName.get(type);
        var objectCreation = String.format("new %s(dbValue)", lazyvalTypeName);
        if(validatedElement.factoryMethod().isPresent()){
            var method = validatedElement.factoryMethod().get();
            objectCreation = String.format("%s.%s(dbValue)", lazyvalTypeName, method.getSimpleName());
        }

        final MethodSpec convertToEntityAttribute = MethodSpec.methodBuilder("convertToEntityAttribute")
                    .addModifiers(Modifier.PUBLIC)
                    .returns(TypeName.get(type))
                    .addParameter(wrappedTypeName, "dbValue")
                    .beginControlFlow("if(dbValue == null)")
                    .addStatement("return null")
                    .endControlFlow()
                    .addStatement("return $L", objectCreation)
                    .build();

        TypeSpec jpaConverter = TypeSpec.classBuilder(element.getSimpleName() + "AttributeConverter")
                .addAnnotation(AnnotationSpec.builder(ClassName.get("jakarta.persistence", "Converter"))
                        .addMember("autoApply", "$L", "true")
                        .build())
                .addSuperinterface(ParameterizedTypeName.get(
                        ClassName.get("jakarta.persistence", "AttributeConverter"),
                        TypeName.get(type),
                        wrappedTypeName))
                .addModifiers(Modifier.PUBLIC)
                .addMethod(convertToDatabaseColumn)
                .addMethod(convertToEntityAttribute)
                .build();

        String jpaConverterPackage = userSettings.get(OPTION_GENERATED_PACKAGE)
                .orElse(String.format("%s.boundary.persistence", extractRootPackage(element)));
        if(jpaConverterPackage.charAt(0) == '.'){
            jpaConverterPackage = jpaConverterPackage.substring(1);
        }
        final JavaFile javaFile = JavaFile.builder(jpaConverterPackage, jpaConverter).build();
        return new GeneratorResult.Java(
                new GeneratorResult.Metadata(javaFile.packageName(), javaFile.typeSpec().name()),
                javaFile.toString());
    }
}

File: ksp/src/main/kotlin/com/qualityminds/lazyval/ksp/codegen/JpaGenerator.kt
package com.qualityminds.lazyval.ksp.codegen

import com.qualityminds.lazyval.ksp.spi.FilePerTypeGenerator
import com.qualityminds.lazyval.ksp.spi.GeneratorResult
import com.qualityminds.lazyval.ksp.spi.SpiGenerator
import com.qualityminds.lazyval.ksp.spi.ValidatedKspGeneratorElement
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName

// tag::docu[]
class JpaGenerator : FilePerTypeGenerator {

    companion object {
        private const val GENERATOR_ID = "jpa"
        private const val OPTION_GENERATED_PACKAGE = "lazyval.jpa.generatedPackage"
    }

    override fun generatorId(): String = GENERATOR_ID

    override fun requiredClasspath(): Collection<String> = listOf("jakarta.persistence.AttributeConverter")

    override fun supportedOptions(): Set<String> {
        return setOf(OPTION_GENERATED_PACKAGE)
    }

    override fun generateFilePerType(
        validatedElement: ValidatedKspGeneratorElement,
        userSettings: SpiGenerator.Settings
    ): GeneratorResult {
        // end::docu[]
        val element = validatedElement.element
        val lazyvalTypeName = element.toClassName()
        val wrappedType = validatedElement.wrappedProperty
        // KotlinPoet specific type-name
        val wrappedTypeName = wrappedType.type.toTypeName()

        val converterClassName = "${validatedElement.typeName}AttributeConverter"

        // Build convertToDatabaseColumn method - use the Kotlin accessor method name
        val convertToDatabaseColumn = FunSpec.builder("convertToDatabaseColumn")
            .addModifiers(KModifier.OVERRIDE)
            .returns(wrappedTypeName.copy(nullable = true))
            .addParameter("type", lazyvalTypeName.copy(nullable = true))
            .apply {
                addStatement("return type?.${validatedElement.kotlinAccessor}")
            }
            .build()

        val parameterName = "dbValue"
        // Build convertToEntityAttribute method
        val convertToEntityAttribute = FunSpec.builder("convertToEntityAttribute")
            .addModifiers(KModifier.OVERRIDE)
            .returns(lazyvalTypeName.copy(nullable = true))
            .addParameter(parameterName, wrappedTypeName.copy(nullable = true))
            .apply {
                addStatement("return $parameterName?.let { ${validatedElement.objectCreation(parameterName)} }")
            }
            .build()

        val converterClass = TypeSpec.classBuilder(converterClassName)
            .addAnnotation(
                AnnotationSpec.builder(ClassName("jakarta.persistence", "Converter"))
                    .addMember("autoApply = true")
                    .build()
            )
            .addSuperinterface(
                ClassName("jakarta.persistence", "AttributeConverter")
                    .parameterizedBy(
                        lazyvalTypeName.copy(nullable = true),
                        wrappedTypeName.copy(nullable = true)
                    )
            )
            .addFunction(convertToDatabaseColumn)
            .addFunction(convertToEntityAttribute)
            .build()

        // Determine package
        val packageName = userSettings.options[OPTION_GENERATED_PACKAGE]
            ?: "${extractRootPackage(element)}.boundary.persistence"

        val fileSpec = FileSpec.builder(packageName, converterClassName).addType(converterClass).build()

        return GeneratorResult.Kotlin(
            GeneratorResult.Metadata(packageName, converterClassName),
            fileSpec.toString())

    }
}